### PR TITLE
Move LB policy instantiation from resolvers into client_channel.

### PR DESCRIPTION
**Pull request by markdroth**
This depends on #7991.

This paves the way for a future PR, which will add service config data to grpc_resolver_result.

### CODE REVIEW

**Comment by dgquintas** [comment]
Prior to actually reviewing, it'd be great if all these changes go in after https://github.com/grpc/grpc/pull/7746 is merged. @ctiller or @markdroth , could you have a last cursory look and LGTM if applicable?

**Comment by markdroth** [comment]
@ctiller, I'd like you to verify that the overall approach is okay here, since this relates to our recent discussions about name resolution semantics.

Once the overall approach is deemed okay, @dgquintas can do the detailed code review.

Thanks!

**Comment by ctiller** [comment]
I thought the outcome of our discussions had been that it's ok for resolvers to call some of these factories directly, however this PR seems to be moving in exactly the opposite direction.

File: src/core/ext/transport/chttp2/client/secure/secure_channel_create.c [hunk 1]
```diff
@@ -276,10 +271,10 @@ static grpc_channel *client_channel_factory_create_channel(
                                               GRPC_CLIENT_CHANNEL, NULL);
   grpc_channel_args_destroy(final_args);
 
-  grpc_resolver *resolver = grpc_resolver_create(target, &f->base);
+  grpc_resolver *resolver = grpc_resolver_create(target);
   if (resolver != NULL) {
-    grpc_client_channel_set_resolver(
-        exec_ctx, grpc_channel_get_channel_stack(channel), resolver);
+    grpc_client_channel_set_resolver_and_client_channel_factory(
```
Comment (ctiller) [inline src/core/ext/transport/chttp2/client/secure/secure_channel_create.c]: What happens when the next factory object needs to be handed in?

grpc_client_channel_set_resolver_and_client_channel_factory_and_load_balancing_factory_and_retry_policy_factory(...);

Comment (markdroth) [inline src/core/ext/transport/chttp2/client/secure/secure_channel_create.c]: If you'd prefer, I can either leave this method with its original name, or I can provide two different methods, one for each thing we need to set.

As an aside, I've actually been thinking that the client channel factory interface is a bit messy and that we should come up with a way to clean it up.  But I don't yet have any concrete thoughts on how to do that.

**Comment by markdroth** [comment]
I've merged in the recent gRPC-LB changes, so this is now ready for review.

**Comment by markdroth** [comment]
This is also part of addressing #7818.

File: src/core/ext/transport/chttp2/client/secure/secure_channel_create.c (hunk 1 above)
Comment (ctiller) [inline src/core/ext/transport/chttp2/client/secure/secure_channel_create.c]: grpc_client_channel_finish_initialization(...)?

Comment (markdroth) [inline src/core/ext/transport/chttp2/client/secure/secure_channel_create.c]: Done.

**Comment by markdroth** [comment]
Thanks, Craig.

David, waiting on a detailed code review from you.

**Comment by dgquintas** [comment]
On it right now :)

File: src/core/ext/client_config/client_channel.c [hunk 2]
```diff
@@ -173,20 +176,26 @@ static void on_resolver_result_changed(grpc_exec_ctx *exec_ctx, void *arg,
   grpc_error *state_error = GRPC_ERROR_CREATE("No load balancing policy");
 
   if (chand->resolver_result != NULL) {
-    lb_policy = grpc_resolver_result_get_lb_policy(chand->resolver_result);
+    grpc_lb_policy_args lb_policy_args;
+    lb_policy_args.addresses =
+        grpc_resolver_result_get_addresses(chand->resolver_result);
```
Comment (dgquintas) [inline src/core/ext/client_config/client_channel.c]: why not inline the function? It seems it's only used here and it's a one-liner. Same for `grpc_resolver_result_get_lb_policy_args` and `grpc_resolver_result_get_lb_policy_name` below.

File: src/core/ext/client_config/client_channel.c [hunk 3]
```diff
@@ -784,6 +797,8 @@ void grpc_client_channel_set_resolver(grpc_exec_ctx *exec_ctx,
     grpc_resolver_next(exec_ctx, resolver, &chand->resolver_result,
                        &chand->on_resolver_result_changed);
   }
+  chand->client_channel_factory = client_channel_factory;
+  grpc_client_channel_factory_ref(client_channel_factory);
```
Comment (dgquintas) [inline src/core/ext/client_config/client_channel.c]: I'd suggest adding an assert over `client_channel_factory` at the beginning of the function, both as "documentation" and also to prevent funky failures if we try to unref a `NULL`.

File: src/core/ext/client_config/lb_policy_factory.c [hunk 4]
```diff
@@ -46,6 +47,23 @@ grpc_lb_addresses* grpc_lb_addresses_create(size_t num_addresses) {
   return addresses;
 }
 
+grpc_lb_addresses* grpc_lb_addresses_copy(grpc_lb_addresses* addresses,
+                                          void* (*user_data_copy)(void*)) {
+  grpc_lb_addresses* new = grpc_lb_addresses_create(addresses->num_addresses);
```
Comment (dgquintas) [inline src/core/ext/client_config/lb_policy_factory.c]: not sure, but using `new` as a variable name may wreak havoc when combining this with C++, being a reserved keyword. `s/new/new_lb_addr/`?

File: src/core/ext/client_config/lb_policy_factory.h [hunk 5]
```diff
@@ -82,9 +87,12 @@ void grpc_lb_addresses_destroy(grpc_lb_addresses *addresses,
                                void (*user_data_destroy)(void *));
 
 /** Arguments passed to LB policies. */
+/* TODO(roth, ctiller): Consider replacing this struct with
+   grpc_channel_args.  See comment in resolver_result.h for details. */
 typedef struct grpc_lb_policy_args {
   grpc_lb_addresses *addresses;
   grpc_client_channel_factory *client_channel_factory;
+  grpc_channel_args *additional_args;
```
Comment (dgquintas) [inline src/core/ext/client_config/lb_policy_factory.h]: Could you add a comment describing why is this needed and what would be an example of an additional arg?

File: src/core/ext/client_config/client_channel.c (hunk 2 above)
Comment (markdroth) [inline src/core/ext/client_config/client_channel.c]: The `grpc_resolver_result` struct is opaque to callers.  Its fields can only be accessed by code in `resolver_result.c`.

File: src/core/ext/client_config/client_channel.c (hunk 3 above)
Comment (markdroth) [inline src/core/ext/client_config/client_channel.c]: Done.

File: src/core/ext/client_config/lb_policy_factory.c (hunk 4 above)
Comment (markdroth) [inline src/core/ext/client_config/lb_policy_factory.c]: Done.

File: src/core/ext/client_config/lb_policy_factory.h (hunk 5 above)
Comment (markdroth) [inline src/core/ext/client_config/lb_policy_factory.h]: I added a comment explaining that it's for passing in implementation-specific parameters to the LB policy.  But since we don't actually have any use-cases for this yet, I don't have an example to add.

File: src/core/ext/client_config/client_channel.c (hunk 2 above)
Comment (dgquintas) [inline src/core/ext/client_config/client_channel.c]: Ah, yes. I got it confused with another one. SG.

**Comment by markdroth** [comment]
The remaining test failures look like flakes, so I'll go ahead and merge this.
